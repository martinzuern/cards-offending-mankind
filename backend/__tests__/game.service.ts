import { v4 as uuidv4 } from 'uuid';
import _ from 'lodash';

import GameService from '../server/services/game.service';
import { InternalGameState, UUID } from '../root-types';

const GAME_PASSWORD = 'super-secure';
let ukGameState: InternalGameState;

describe('GameService', () => {
  describe('static information', () => {
    it('lists available packs', () => {
      const packs = GameService.getAvailablePacks();
      expect(packs).toMatchSnapshot();
    });
  });

  describe('create / join game', () => {
    it('creates empty game', async (done) => {
      const gameState = await GameService.initGameState({});
      expect(gameState).toMatchSnapshot({
        game: {
          id: expect.any(String),
        },
      });
      done();
    });

    it('creates game with two packs and password', async (done) => {
      const gameState = await GameService.initGameState({
        password: GAME_PASSWORD,
        packs: [{ abbr: 'BaseUK' }, { abbr: 'www' }],
      });
      expect(gameState).toMatchSnapshot({
        game: {
          id: expect.any(String),
          password: expect.any(String),
        },
      });
      ukGameState = gameState;
      done();
    });

    it('rejects invalid packs', async (done) => {
      expect.assertions(2);
      try {
        await GameService.initGameState({
          password: GAME_PASSWORD,
          // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
          // @ts-ignore
          packs: [{ abbr: 'BaseUK' }, {}],
        });
      } catch (error) {
        expect(error).toMatchSnapshot();
      }
      try {
        await GameService.initGameState({
          password: GAME_PASSWORD,
          packs: [{ abbr: 'BaseUKf' }],
        });
      } catch (error) {
        expect(error).toMatchSnapshot();
      }
      done();
    });

    describe('validates password', () => {
      let emptyGameState;

      beforeAll(async (done) => {
        emptyGameState = await GameService.initGameState({});
        done();
      });

      it('accepts correct', async (done) => {
        const result = await Promise.all([
          GameService.validateGamePassword(emptyGameState.game, ''),
          GameService.validateGamePassword(emptyGameState.game, 'foo'),
          GameService.validateGamePassword(ukGameState.game, GAME_PASSWORD),
        ]);
        expect(result).toMatchInlineSnapshot(`
          Array [
            true,
            true,
            true,
          ]
        `);
        done();
      });
      it('rejects undefined', async (done) => {
        expect.assertions(2);
        try {
          await GameService.validateGamePassword(ukGameState.game, undefined);
        } catch (error) {
          expect(error).toMatchInlineSnapshot(
            `[AssertionError: Error: Game password not provided.]`
          );
        }

        try {
          await GameService.validateGamePassword(ukGameState.game, '');
        } catch (error) {
          expect(error).toMatchInlineSnapshot(`[AssertionError: Error: Game password incorrect]`);
        }
        done();
      });
      it('rejects wrong type', async (done) => {
        expect.assertions(2);
        try {
          await GameService.validateGamePassword(ukGameState.game, (true as unknown) as string);
        } catch (error) {
          expect(error).toMatchInlineSnapshot(
            `[AssertionError: Error: Game password not provided.]`
          );
        }

        try {
          await GameService.validateGamePassword(ukGameState.game, (22 as unknown) as string);
        } catch (error) {
          expect(error).toMatchInlineSnapshot(
            `[AssertionError: Error: Game password not provided.]`
          );
        }
        done();
      });
      it('rejects incorrect', async (done) => {
        expect.assertions(1);
        try {
          await GameService.validateGamePassword(ukGameState.game, 'foo');
        } catch (error) {
          expect(error).toMatchInlineSnapshot(`[AssertionError: Error: Game password incorrect]`);
        }
        done();
      });
    });
  });

  describe('add player', () => {
    const gameId = uuidv4() as UUID;

    it('throws on missing nickname', () => {
      expect.assertions(1);
      try {
        GameService.initPlayer(gameId, {});
      } catch (error) {
        expect(error).toMatchInlineSnapshot(`[AssertionError: undefined == true]`);
      }
    });

    it('returns valid player', () => {
      const player = GameService.initPlayer(gameId, { nickname: 'foo' });
      expect(player).toMatchSnapshot({
        id: expect.any(String),
        token: expect.any(String),
      });
    });
  });

  describe('builds pile', () => {
    it('works', () => {
      const pile = GameService.buildPile(ukGameState.game);
      expect(pile.prompts.length).toBe(99);
      expect(pile.responses.length).toBe(481);
      expect({
        ...pile,
        prompts: expect.any(Array),
        responses: expect.any(Array),
      }).toMatchSnapshot();
    });

    it('is shuffled', () => {
      const pile1 = GameService.buildPile(ukGameState.game);
      const pile2 = GameService.buildPile(ukGameState.game);
      expect(pile1).not.toEqual(pile2);
    });
  });

  describe('start game', () => {
    it('works', async (done) => {
      const baseGameState = await GameService.initGameState({
        packs: [{ abbr: 'BaseUK' }],
      });
      baseGameState.players.push(
        _.omit(GameService.initPlayer('', { nickname: 'foo1', isActive: true }), 'token'),
        _.omit(GameService.initPlayer('', { nickname: 'foo2', isActive: true }), 'token'),
        _.omit(GameService.initPlayer('', { nickname: 'foo3', isActive: true }), 'token')
      );
      const newGameState = GameService.startGame(baseGameState);
      expect(newGameState.players.length).toBe(3);
      expect(newGameState.players.every((p) => p.deck.length === 10)).toBe(true);
      expect(newGameState.piles.responses.length).toBe(460 - 30);
      expect(newGameState.piles.prompts.length).toBe(90 - 1);
      done();
    });
  });
});
