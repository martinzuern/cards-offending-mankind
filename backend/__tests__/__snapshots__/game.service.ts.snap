// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameService add player returns valid player 1`] = `
Object {
  "deck": Array [],
  "id": Any<String>,
  "isAI": false,
  "isActive": false,
  "isHost": false,
  "nickname": "foo",
  "points": 0,
  "token": Any<String>,
}
`;

exports[`GameService builds pile works 1`] = `
Object {
  "discardedPrompts": Array [],
  "discardedResponses": Array [],
  "prompts": Any<Array>,
  "responses": Any<Array>,
}
`;

exports[`GameService create / join game creates empty game 1`] = `
Object {
  "game": Object {
    "handSize": 10,
    "hasPassword": false,
    "id": Any<String>,
    "packs": Array [],
    "status": "created",
    "timeouts": Object {
      "betweenRounds": 30,
      "judging": 120,
      "playing": 120,
      "revealing": 60,
    },
    "winnerPoints": 20,
  },
  "players": Array [],
  "rounds": Array [],
}
`;

exports[`GameService create / join game creates game with two packs and password 1`] = `
Object {
  "game": Object {
    "handSize": 10,
    "hasPassword": true,
    "id": Any<String>,
    "packs": Array [
      Object {
        "abbr": "BaseUK",
        "description": "The UK edition of the base game",
        "icon": "",
        "name": "Base Set (UK Edition)",
        "official": true,
      },
      Object {
        "abbr": "www",
        "description": "- placeholder -",
        "icon": "globe",
        "name": "World Wide Web Pack",
        "official": true,
      },
    ],
    "password": Any<String>,
    "status": "created",
    "timeouts": Object {
      "betweenRounds": 30,
      "judging": 120,
      "playing": 120,
      "revealing": 60,
    },
    "winnerPoints": 20,
  },
  "players": Array [],
  "rounds": Array [],
}
`;

exports[`GameService create / join game rejects invalid packs 1`] = `[AssertionError: Invalid pack undefined.]`;

exports[`GameService create / join game rejects invalid packs 2`] = `[AssertionError: Invalid pack BaseUKf.]`;

exports[`GameService static information lists available packs 1`] = `
Array [
  Object {
    "abbr": "Base",
    "description": "- placeholder -",
    "icon": "",
    "name": "Base Set",
    "official": true,
    "promptsCount": 90,
    "responsesCount": 460,
  },
  Object {
    "abbr": "BaseUK",
    "description": "The UK edition of the base game",
    "icon": "",
    "name": "Base Set (UK Edition)",
    "official": true,
    "promptsCount": 90,
    "responsesCount": 460,
  },
  Object {
    "abbr": "family",
    "description": "Cards Against Humanity: Family Edition is a new fill-in-the-blank comedy game that's just like Cards Against Humanity, but it's written for kids and adults to play together.",
    "icon": "child",
    "name": "Family Edition (BETA)",
    "official": true,
    "promptsCount": 70,
    "responsesCount": 349,
  },
  Object {
    "abbr": "greenbox",
    "description": "- placeholder -",
    "icon": "square",
    "name": "Green Box Expansion",
    "official": true,
    "promptsCount": 55,
    "responsesCount": 245,
  },
  Object {
    "abbr": "CAHe1",
    "description": "- placeholder -",
    "icon": "1",
    "name": "The First Expansion",
    "official": true,
    "promptsCount": 20,
    "responsesCount": 80,
  },
  Object {
    "abbr": "CAHe2",
    "description": "- placeholder -",
    "icon": "2",
    "name": "The Second Expansion",
    "official": true,
    "promptsCount": 25,
    "responsesCount": 75,
  },
  Object {
    "abbr": "CAHe3",
    "description": "- placeholder -",
    "icon": "3",
    "name": "The Third Expansion",
    "official": true,
    "promptsCount": 25,
    "responsesCount": 75,
  },
  Object {
    "abbr": "CAHe6",
    "description": "- placeholder -",
    "icon": "6",
    "name": "The Sixth Expansion",
    "official": true,
    "promptsCount": 25,
    "responsesCount": 75,
  },
  Object {
    "abbr": "CAHe5",
    "description": "- placeholder -",
    "icon": "5",
    "name": "The Fifth Expansion",
    "official": true,
    "promptsCount": 25,
    "responsesCount": 75,
  },
  Object {
    "abbr": "CAHe4",
    "description": "- placeholder -",
    "icon": "4",
    "name": "The Fourth Expansion",
    "official": true,
    "promptsCount": 30,
    "responsesCount": 70,
  },
  Object {
    "abbr": "PAXP2013",
    "description": "- placeholder -",
    "icon": "paypal",
    "name": "PAX Prime 2013",
    "official": true,
    "promptsCount": 7,
    "responsesCount": 37,
  },
  Object {
    "abbr": "BF2019AI",
    "description": "30 cards actually written by a machine learning algorithm trained on our brainstorming documents.",
    "icon": "robot",
    "name": "AI pack",
    "official": true,
    "promptsCount": 0,
    "responsesCount": 30,
  },
  Object {
    "abbr": "BF2019Human",
    "description": "30 beautiful cards about the human condition straight from the hearts of our human writers.",
    "icon": "user",
    "name": "Human pack",
    "official": true,
    "promptsCount": 2,
    "responsesCount": 28,
  },
  Object {
    "abbr": "cats",
    "description": "AT TONIGHT'S PERFORMANCE OF: The Cards Against Humanity Theatre Pack THE ROLE OF: 30 relatable cards about theatre WILL BE PLAYED BY: 30 cards about the musical CATS. We apologize for the inconvenience. Please accept this $5 refund.",
    "icon": "cat",
    "name": "Cats Pack",
    "official": true,
    "promptsCount": 3,
    "responsesCount": 27,
  },
  Object {
    "abbr": "Ass",
    "description": "Back it up. Now bring it down. Lower. Down to the floor. Now wiggle it all around. Now get out your wallet and buy the all-new Ass Pack! 30 new cards regarding butts. Includes cards by visionary ass connoisseur Sir Mix-A-Lot. Round. Juicy. Spank it!",
    "icon": "wind",
    "name": "Ass Pack",
    "official": true,
    "promptsCount": 3,
    "responsesCount": 27,
  },
  Object {
    "abbr": "fantasy",
    "description": "- placeholder -",
    "icon": "magic",
    "name": "Fantasy Pack",
    "official": true,
    "promptsCount": 6,
    "responsesCount": 26,
  },
  Object {
    "abbr": "pride",
    "description": "It’s Pride Month, the time of year when companies put a rainbow on their products to capitalize on gay culture. Fabulous! 30 cards that completely encapsulate the queer experience. Written by an elite squad of lesbians, bisexuals, gays, trans people, allies, and unicorns. All profits donated to Howard Brown Health.",
    "icon": "rainbow",
    "name": "Pride Pack",
    "official": true,
    "promptsCount": 5,
    "responsesCount": 25,
  },
  Object {
    "abbr": "00s",
    "description": "Dude, where’s my car? The 2000s Nostalgia Pack comes with 30 new cards capitalizing on the surge of dopamine that floods your brain when you think about your childhood. Chunky highlights. Butterfly clips. Bootcut jeans. Bush did 9/11. You’re such a Miranda!",
    "icon": "backward",
    "name": "2000s Nostalgia Pack",
    "official": true,
    "promptsCount": 5,
    "responsesCount": 25,
  },
  Object {
    "abbr": "theatre",
    "description": "As the Bard once said, “All the world’s a stage, and all the men and women merely members of our target demographic.” Hark! It’s the Theatre Pack, 30 cards about those things that are like movies but they happen right in front of you. We’d like to thank our parents, our amazing director, and of course, God. Lin-Manuel Miranda’s favorite pack. All profits fund a grant program for innovative theater and comedy run by the Theater Communications Group.",
    "icon": "theater-masks",
    "name": "Theatre Pack",
    "official": true,
    "promptsCount": 5,
    "responsesCount": 25,
  },
  Object {
    "abbr": "PAXE2013",
    "description": "- placeholder -",
    "icon": "arrow-circle-right",
    "name": "PAX East 2013",
    "official": true,
    "promptsCount": 6,
    "responsesCount": 24,
  },
  Object {
    "abbr": "dad",
    "description": "Surprise! This DVD case is actually the Cards Against Humanity Dad Pack, and fatherhood is actually a trap that lures virile men into lives of tinkering with thermostats and shopping for bargain DVDs at Target. The Dad Pack includes 30 totally new cards about the joys of fatherhood. Happy Father’s Day!",
    "icon": "user-tie",
    "name": "Dad Pack",
    "official": true,
    "promptsCount": 6,
    "responsesCount": 24,
  },
  Object {
    "abbr": "food",
    "description": "- placeholder -",
    "icon": "cutlery",
    "name": "Food Pack",
    "official": true,
    "promptsCount": 6,
    "responsesCount": 24,
  },
  Object {
    "abbr": "geek",
    "description": "30 cards about video games, D&D, Game of Thrones, and all the other bullshit you like. Previously released in packs given away at PAX East and PAX Prime in 2013 and 2014.",
    "icon": "glasses",
    "name": "Geek Pack",
    "official": true,
    "promptsCount": 5,
    "responsesCount": 24,
  },
  Object {
    "abbr": "PAX2015",
    "description": "- placeholder -",
    "icon": "truck",
    "name": "PAX Prime 2015 Food Packs",
    "official": true,
    "promptsCount": 6,
    "responsesCount": 24,
  },
  Object {
    "abbr": "college",
    "description": "Remember when you were in college and ate ramen noodles and had sex all the time? Wouldn’t it be great to relive those glory days? 30 brand-new cards about learning and throw up. Comes with a special 18″ x 24″ poster for your dorm room.",
    "icon": "graduation-cap",
    "name": "College Pack",
    "official": true,
    "promptsCount": 6,
    "responsesCount": 24,
  },
  Object {
    "abbr": "reject2",
    "description": "- placeholder -",
    "icon": "ban",
    "name": "Reject Pack 2",
    "official": true,
    "promptsCount": 10,
    "responsesCount": 24,
  },
  Object {
    "abbr": "90s",
    "description": "- placeholder -",
    "icon": "birthday-cake",
    "name": "90s Nostalgia Pack",
    "official": true,
    "promptsCount": 7,
    "responsesCount": 23,
  },
  Object {
    "abbr": "xmas2012",
    "description": "- placeholder -",
    "icon": "tree",
    "name": "2012 Holiday Pack",
    "official": true,
    "promptsCount": 7,
    "responsesCount": 23,
  },
  Object {
    "abbr": "save",
    "description": "Remember that time when we all thought the Trump presidency was going to be an unmitigated disaster? And then Cards Against Humanity saved America? Well, we still have the cards from that. Contains 30 cards previously released as part of our 2017 Cards Against Humanity Saves America campaign. Whoops, Donald Trump is still the president. Oh god.",
    "icon": "flag-usa",
    "name": "Cards Against Humanity Saves America Pack",
    "official": true,
    "promptsCount": 7,
    "responsesCount": 23,
  },
  Object {
    "abbr": "reject3",
    "description": "25 cards (23 white, 2 black) that were previously withheld from being released for one reason or another. Each co-creator picked three cards that were rejected from print.",
    "icon": "times",
    "name": "Reject Pack 3",
    "official": true,
    "promptsCount": 2,
    "responsesCount": 23,
  },
  Object {
    "abbr": "science",
    "description": "- placeholder -",
    "icon": "flask",
    "name": "Science Pack",
    "official": true,
    "promptsCount": 7,
    "responsesCount": 23,
  },
  Object {
    "abbr": "PAXE2014",
    "description": "- placeholder -",
    "icon": "arrow-circle-o-right",
    "name": "PAX East 2014",
    "official": true,
    "promptsCount": 5,
    "responsesCount": 22,
  },
  Object {
    "abbr": "trumpbag",
    "description": "- placeholder -",
    "icon": "medkit",
    "name": "Trump Survival Pack",
    "official": true,
    "promptsCount": 3,
    "responsesCount": 22,
  },
  Object {
    "abbr": "www",
    "description": "- placeholder -",
    "icon": "globe",
    "name": "World Wide Web Pack",
    "official": true,
    "promptsCount": 9,
    "responsesCount": 21,
  },
  Object {
    "abbr": "Box",
    "description": "- placeholder -",
    "icon": "cube",
    "name": "Box Expansion",
    "official": true,
    "promptsCount": 0,
    "responsesCount": 21,
  },
  Object {
    "abbr": "Canadian",
    "description": "- placeholder -",
    "icon": "leaf",
    "name": "Canadian",
    "official": true,
    "promptsCount": 5,
    "responsesCount": 21,
  },
  Object {
    "abbr": "xmas2013",
    "description": "- placeholder -",
    "icon": "gift",
    "name": "2013 Holiday Pack",
    "official": true,
    "promptsCount": 9,
    "responsesCount": 21,
  },
  Object {
    "abbr": "retailtarget",
    "description": "A 20 card expansion Pack for Cards Againt Humanity, available only at Target stores.",
    "icon": "bullseye",
    "name": "Target Retail Product Pack",
    "official": true,
    "promptsCount": 1,
    "responsesCount": 19,
  },
  Object {
    "abbr": "desertbus",
    "description": "- placeholder -",
    "icon": "bus",
    "name": "Desert Bus For Hope Pack",
    "official": true,
    "promptsCount": 5,
    "responsesCount": 19,
  },
  Object {
    "abbr": "xmas2014",
    "description": "- placeholder -",
    "icon": "sleigh",
    "name": "2014 Holiday Pack",
    "official": true,
    "promptsCount": 5,
    "responsesCount": 18,
  },
  Object {
    "abbr": "midterm2018",
    "description": "All-new Midterm Pack with 18 cards about the surreal fever dream that is American politics in 2018. All profits will be donated to Run for Something.",
    "icon": "vote-yea",
    "name": "Midterm Pack 2018",
    "official": true,
    "promptsCount": 2,
    "responsesCount": 16,
  },
  Object {
    "abbr": "HOCAH",
    "description": "- placeholder -",
    "icon": "university",
    "name": "House of Cards Against Humanity",
    "official": true,
    "promptsCount": 9,
    "responsesCount": 16,
  },
  Object {
    "abbr": "reject",
    "description": "- placeholder -",
    "icon": "thumbs-down",
    "name": "Reject Pack",
    "official": true,
    "promptsCount": 8,
    "responsesCount": 16,
  },
  Object {
    "abbr": "hidden",
    "description": "A pack hidden in a hidden compartment in a combined Exploding Kittens and Cards Against Humanity end cap display in certain Targets.",
    "icon": "eye",
    "name": "Hidden Compartment Pack",
    "official": true,
    "promptsCount": 0,
    "responsesCount": 15,
  },
  Object {
    "abbr": "fascism",
    "description": "- placeholder -",
    "icon": "skull",
    "name": "Fascism Pack",
    "official": true,
    "promptsCount": 2,
    "responsesCount": 15,
  },
  Object {
    "abbr": "jew",
    "description": "Fun fact: 100% of the Cards Against Humanity writers are Jewish. Can you believe it? A Jewish comedy writer! Anything is possible in 2017.",
    "icon": "star-of-david",
    "name": "Jew Pack",
    "official": true,
    "promptsCount": 2,
    "responsesCount": 13,
  },
  Object {
    "abbr": "table",
    "description": "Promo pack bonus for all of those who backed $20 or more for Season 3 of Tabletop Kickstarter campaign.",
    "icon": "dice-d20",
    "name": "Tabletop Pack",
    "official": true,
    "promptsCount": 3,
    "responsesCount": 12,
  },
  Object {
    "abbr": "trumpvote",
    "description": "- placeholder -",
    "icon": "bullhorn",
    "name": "Vote for Trump Pack",
    "official": true,
    "promptsCount": 3,
    "responsesCount": 12,
  },
  Object {
    "abbr": "hillary",
    "description": "- placeholder -",
    "icon": "venus",
    "name": "Vote for Hillary Pack",
    "official": true,
    "promptsCount": 3,
    "responsesCount": 11,
  },
  Object {
    "abbr": "Seasons",
    "description": "A Target exclusive holiday pack containing 12 new cards (2 prompt cards, 10 response cards) and 18 cards from previous holiday packs.",
    "icon": "snowflake",
    "name": "Non Denominational Seasons Greetings Pack",
    "official": true,
    "promptsCount": 2,
    "responsesCount": 10,
  },
  Object {
    "abbr": "masseffect",
    "description": "A limited-edition pack made in collaboration with Mass Effect.",
    "icon": "meteor",
    "name": "Mass Effect Pack",
    "official": true,
    "promptsCount": 4,
    "responsesCount": 10,
  },
  Object {
    "abbr": "misprint",
    "description": "- placeholder -",
    "icon": "print",
    "name": "Misprint Replacement Bonus Cards",
    "official": true,
    "promptsCount": 1,
    "responsesCount": 9,
  },
  Object {
    "abbr": "PAXEP2014",
    "description": "- placeholder -",
    "icon": "users",
    "name": "PAX East 2014 Panel Pack",
    "official": true,
    "promptsCount": 2,
    "responsesCount": 8,
  },
  Object {
    "abbr": "Hanukkah",
    "description": "A top secret 7 card pack only obtainable through the 2015 7 gifts of Hanukkah promotion. Only three packs were printed.",
    "icon": "hanukiah",
    "name": "Hanukkah LOL Pack",
    "official": true,
    "promptsCount": 0,
    "responsesCount": 7,
  },
  Object {
    "abbr": "Gems",
    "description": "The Hidden Gems Bundle is an expansion pack to Cards Against Humanity. Contains six different themed packs you forgot to buy when they came out. Also includes a few all-new cards we desperately crammed in at the last second. An unbelievable deal. Amazon exclusive.",
    "icon": "gem",
    "name": "Hidden Gems Bundle",
    "official": true,
    "promptsCount": 3,
    "responsesCount": 7,
  },
  Object {
    "abbr": "jackwhite",
    "description": "- placeholder -",
    "icon": "guitar",
    "name": "Jack White Show Pack",
    "official": true,
    "promptsCount": 1,
    "responsesCount": 6,
  },
  Object {
    "abbr": "PAXPP2014",
    "description": "- placeholder -",
    "icon": "user-times",
    "name": "PAX Prime 2014 Panel Pack",
    "official": true,
    "promptsCount": 5,
    "responsesCount": 6,
  },
  Object {
    "abbr": "retail",
    "description": "Cards Against Humanity: Retail Pack is a 5-card expansion for Cards Against Humanity. It is included with a purchase of the base game from a physical store that was accepted into Cards Against Humanity's retailer program.",
    "icon": "store-alt",
    "name": "Retail Pack",
    "official": true,
    "promptsCount": 2,
    "responsesCount": 3,
  },
  Object {
    "abbr": "blackbox",
    "description": "Released as part of the Blackbox Press Kit.",
    "icon": "th-large",
    "name": "Blackbox Press Kit",
    "official": true,
    "promptsCount": 0,
    "responsesCount": 1,
  },
  Object {
    "abbr": "Weed",
    "description": "- placeholder -",
    "icon": "pagelines",
    "name": "Weed Pack",
    "official": true,
    "promptsCount": 0,
    "responsesCount": 0,
  },
  Object {
    "abbr": "Period",
    "description": "- placeholder -",
    "icon": "tint",
    "name": "Period Pack",
    "official": true,
    "promptsCount": 0,
    "responsesCount": 0,
  },
  Object {
    "abbr": "apples",
    "description": "- placeholder -",
    "icon": "apple",
    "name": "[$] Apples to Apples&reg; Party Pack",
    "official": false,
    "promptsCount": 249,
    "responsesCount": 747,
  },
  Object {
    "abbr": "c-anime",
    "description": "- placeholder -",
    "icon": "sun-o",
    "name": "[C] Anime",
    "official": false,
    "promptsCount": 151,
    "responsesCount": 525,
  },
  Object {
    "abbr": "c-guywglasses",
    "description": "- placeholder -",
    "icon": "street-view",
    "name": "[C] That Guy With The Glasses",
    "official": false,
    "promptsCount": 207,
    "responsesCount": 424,
  },
  Object {
    "abbr": "c-homestuck",
    "description": "- placeholder -",
    "icon": "home",
    "name": "[C] Cards Against Homestuck",
    "official": false,
    "promptsCount": 67,
    "responsesCount": 340,
  },
  Object {
    "abbr": "crabs",
    "description": "- placeholder -",
    "icon": "wrench",
    "name": "[$] Crabs Adjust Humidity",
    "official": false,
    "promptsCount": 129,
    "responsesCount": 320,
  },
  Object {
    "abbr": "HACK",
    "description": "- placeholder -",
    "icon": "laptop",
    "name": "[C] Hackers Against Humanity",
    "official": false,
    "promptsCount": 78,
    "responsesCount": 279,
  },
  Object {
    "abbr": "c-tg",
    "description": "- placeholder -",
    "icon": "reddit",
    "name": "[C] /tg/",
    "official": false,
    "promptsCount": 90,
    "responsesCount": 266,
  },
  Object {
    "abbr": "matrimony",
    "description": "- placeholder -",
    "icon": "bell",
    "name": "[$] Cads About Matrimony",
    "official": false,
    "promptsCount": 55,
    "responsesCount": 230,
  },
  Object {
    "abbr": "c-admin",
    "description": "- placeholder -",
    "icon": "thumbs-o-up",
    "name": "[C] Admin's Picks",
    "official": false,
    "promptsCount": 123,
    "responsesCount": 227,
  },
  Object {
    "abbr": "c-vidya",
    "description": "- placeholder -",
    "icon": "video-camera",
    "name": "[C] Vidya",
    "official": false,
    "promptsCount": 107,
    "responsesCount": 215,
  },
  Object {
    "abbr": "c-sodomydog",
    "description": "- placeholder -",
    "icon": "paw",
    "name": "[C] Sodomy Dog's Furry Pack",
    "official": false,
    "promptsCount": 105,
    "responsesCount": 178,
  },
  Object {
    "abbr": "c-gamegrumps",
    "description": "- placeholder -",
    "icon": "frown-o",
    "name": "[C] Game Grumps",
    "official": false,
    "promptsCount": 99,
    "responsesCount": 178,
  },
  Object {
    "abbr": "c-northernlion",
    "description": "- placeholder -",
    "icon": "youtube-play",
    "name": "[C] Northernlion",
    "official": false,
    "promptsCount": 101,
    "responsesCount": 171,
  },
  Object {
    "abbr": "c-rpanons",
    "description": "- placeholder -",
    "icon": "user-secret",
    "name": "[C] RPAnons",
    "official": false,
    "promptsCount": 69,
    "responsesCount": 163,
  },
  Object {
    "abbr": "c-rt",
    "description": "- placeholder -",
    "icon": "youtube",
    "name": "[C] Rooster Teeth",
    "official": false,
    "promptsCount": 69,
    "responsesCount": 155,
  },
  Object {
    "abbr": "c-equinity",
    "description": "- placeholder -",
    "icon": "trophy",
    "name": "[C] Cards Against Equinity",
    "official": false,
    "promptsCount": 60,
    "responsesCount": 138,
  },
  Object {
    "abbr": "c-stupid",
    "description": "- placeholder -",
    "icon": "lemon-o",
    "name": "[C] Ridiculously Stupid",
    "official": false,
    "promptsCount": 47,
    "responsesCount": 115,
  },
  Object {
    "abbr": "c-vewysewious",
    "description": "- placeholder -",
    "icon": "black-tie",
    "name": "[C] Very Serious",
    "official": false,
    "promptsCount": 59,
    "responsesCount": 113,
  },
  Object {
    "abbr": "c-golby",
    "description": "- placeholder -",
    "icon": "file-text-o",
    "name": "[C] Golby Fan Club",
    "official": false,
    "promptsCount": 18,
    "responsesCount": 101,
  },
  Object {
    "abbr": "NSFH",
    "description": "- placeholder -",
    "icon": "exclamation-triangle",
    "name": "[C] Not Safe For Humanity",
    "official": false,
    "promptsCount": 32,
    "responsesCount": 99,
  },
  Object {
    "abbr": "c-prtg",
    "description": "A free marketing deck from Paessler labeled 'The Game Against Downtime: A card game for IT lovers', each card is branded with 'PRTG Network Monitor'.",
    "icon": "cloudscale",
    "name": "[C] PRTG",
    "official": false,
    "promptsCount": 20,
    "responsesCount": 99,
  },
  Object {
    "abbr": "c-mrman",
    "description": "- placeholder -",
    "icon": "male",
    "name": "[C] Mr. Man Collection",
    "official": false,
    "promptsCount": 37,
    "responsesCount": 99,
  },
  Object {
    "abbr": "c-khaos",
    "description": "- placeholder -",
    "icon": "ils",
    "name": "[C] Khaos WolfKat",
    "official": false,
    "promptsCount": 76,
    "responsesCount": 90,
  },
  Object {
    "abbr": "c-ladies",
    "description": "- placeholder -",
    "icon": "female",
    "name": "[C] Ladies Against Humanity",
    "official": false,
    "promptsCount": 25,
    "responsesCount": 88,
  },
  Object {
    "abbr": "CAHgrognards",
    "description": "- placeholder -",
    "icon": "moon-o",
    "name": "[C] Grognards Against Humanity (RPG fandom pack)",
    "official": false,
    "promptsCount": 22,
    "responsesCount": 86,
  },
  Object {
    "abbr": "GOT",
    "description": "- placeholder -",
    "icon": "shield",
    "name": "[C] Game of Thrones (no spoilers)",
    "official": false,
    "promptsCount": 20,
    "responsesCount": 80,
  },
  Object {
    "abbr": "c-fim",
    "description": "- placeholder -",
    "icon": "heart",
    "name": "[C] FiMFiction.net",
    "official": false,
    "promptsCount": 36,
    "responsesCount": 77,
  },
  Object {
    "abbr": "c-doctorwho",
    "description": "- placeholder -",
    "icon": "stethoscope",
    "name": "[C] Doctor Who",
    "official": false,
    "promptsCount": 26,
    "responsesCount": 71,
  },
  Object {
    "abbr": "c-ragingpsyfag",
    "description": "- placeholder -",
    "icon": "fighter-jet",
    "name": "[C] RagingPsyfag's Pack of Shenanigans",
    "official": false,
    "promptsCount": 16,
    "responsesCount": 62,
  },
  Object {
    "abbr": "c-imgur",
    "description": "- placeholder -",
    "icon": "info",
    "name": "[C] Imgur",
    "official": false,
    "promptsCount": 56,
    "responsesCount": 49,
  },
  Object {
    "abbr": "c-xkcd",
    "description": "- placeholder -",
    "icon": "close",
    "name": "[C] xkcd",
    "official": false,
    "promptsCount": 67,
    "responsesCount": 41,
  },
  Object {
    "abbr": "c-derps",
    "description": "- placeholder -",
    "icon": "meh-o",
    "name": "[C] Derps Against Humanity",
    "official": false,
    "promptsCount": 42,
    "responsesCount": 37,
  },
  Object {
    "abbr": "Image1",
    "description": "- placeholder -",
    "icon": "picture-o",
    "name": "[C] Joey Image 1",
    "official": false,
    "promptsCount": 15,
    "responsesCount": 35,
  },
  Object {
    "abbr": "c-neindy",
    "description": "- placeholder -",
    "icon": "question",
    "name": "[C] NEIndy",
    "official": false,
    "promptsCount": 16,
    "responsesCount": 30,
  },
  Object {
    "abbr": "c-antisocial",
    "description": "- placeholder -",
    "icon": "gavel",
    "name": "[C] Antisocial Injustice",
    "official": false,
    "promptsCount": 9,
    "responsesCount": 27,
  },
  Object {
    "abbr": "c-nobilis",
    "description": "- placeholder -",
    "icon": "exclamation",
    "name": "[C] Nobilis Reed",
    "official": false,
    "promptsCount": 10,
    "responsesCount": 24,
  },
  Object {
    "abbr": "c-socialgamer",
    "description": "- placeholder -",
    "icon": "gamepad",
    "name": "[C] SocialGamer",
    "official": false,
    "promptsCount": 13,
    "responsesCount": 22,
  },
  Object {
    "abbr": "c-eurovision",
    "description": "- placeholder -",
    "icon": "music",
    "name": "[C] Eurovision Song Contest",
    "official": false,
    "promptsCount": 20,
    "responsesCount": 20,
  },
]
`;
